server/bleep/src/semantic/embedder.rs

```diff
@@ -65,89 +59,261 @@ pub struct EmbedChunk {
-        let token_type_ids = tokenizer_output.get_type_ids();
-        let length = input_ids.len();
-        trace!("embedding {} tokens {:?}", length, sequence);
-
-        let inputs_ids_array = ndarray::Array::from_shape_vec(
-            (1, length),
-            input_ids.iter().map(|&x| x as i64).collect(),
-        )?;
-
-        let attention_mask_array = ndarray::Array::from_shape_vec(
-            (1, length),
-            attention_mask.iter().map(|&x| x as i64).collect(),
-        )?;
-
-        let token_type_ids_array = ndarray::Array::from_shape_vec(
-            (1, length),
-            token_type_ids.iter().map(|&x| x as i64).collect(),
-        )?;
-
-        let outputs = self.session.run([
-            InputTensor::from_array(inputs_ids_array.into_dyn()),
-            InputTensor::from_array(attention_mask_array.into_dyn()),
-            InputTensor::from_array(token_type_ids_array.into_dyn()),
-        ])?;
-
-        let output_tensor: OrtOwnedTensor<f32, _> = outputs[0].try_extract().unwrap();
-        let sequence_embedding = &*output_tensor.view();
-        let pooled = sequence_embedding.mean_axis(Axis(1)).unwrap();
-        Ok(pooled.to_owned().as_slice().unwrap().to_vec())
+#[cfg(feature = "metal")]
+mod gpu {
+    use super::*;
+    use tracing::{error, info};
+
+    pub struct LocalEmbedder {
+        model: Box<dyn llm::Model>,
+        sessions: Vec<Arc<tokio::sync::Mutex<llm::InferenceSession>>>,
+        tokenizer: Tokenizer,
+        permits: Arc<tokio::sync::Semaphore>,
     }
 
-    fn tokenizer(&self) -> &Tokenizer {
-        &self.tokenizer
+    // InferenceSession is explicitly not Sync because it uses ggml::Tensor internally,
+    // Bert does not make use of these tensors however
+    unsafe impl Sync for LocalEmbedder {}
+
+    impl LocalEmbedder {
+        pub fn new(model_dir: &Path) -> anyhow::Result<Self> {
+            let mut model_params = llm::ModelParameters::default();
+
+            if cfg!(feature = "metal") {
+                model_params.use_gpu = true; // TODO: make configurable
+            }
+
+            let model = llm::load_dynamic(
+                Some(llm::ModelArchitecture::Bert),
+                &model_dir.join("ggml").join("ggml-model-q4_0.bin"),
+                // this tokenizer is used for embedding
+                llm::TokenizerSource::HuggingFaceTokenizerFile(
+                    model_dir.join("ggml").join("tokenizer.json"),
+                ),
+                model_params,
+                llm::load_progress_callback_stdout,
+            )?;
+
+            // TODO: this can be parameterized
+            //
+            // the lower this number, the more time we might spend waiting to run an embedding.
+            // the higher this number, the more vram we use, currently we use ~2G per session. this
+            // can be fixed by disabling scratch buffers in ggml, bert has no use for this.
+            let session_count = 3;
```

Comment:
the difference in inference speed with a lower session count is minor, we see a ~6% speedup when indexing bloop with `3` instead of `1`.